### ISSUE TITLE
no-unused-vars doesn't count calling a method from a closure in right-hand side

### ISSUE THREAD

**Issue by dgreensp** [denoland/deno_lint#1318]
### Lint Name

no-unused-vars

### Code Snippet

```ts
const foo = new MyClass(() => {
  foo.hello();
});
```

### Expected Result

No lint errors

### Actual Result

Error on `const foo`

### Additional Info

This situation arises when `new MyClass` has a side-effect, for example attaching event handlers.  For the sake of illustration:

```ts
class MyClass {
  constructor(run: () => void) {
    setTimeout(run, 1000);
  }
  hello() {
    console.log("hello");
  }
}
```

In searching for this issue, I only found #667.

### Version

```
deno 1.46.2 (stable, release, aarch64-apple-darwin)
v8 12.9.202.5-rusty
typescript 5.5.2
```
